src/dbus_api/types.rs

```diff
@@ -127,9 +127,20 @@ impl ActionQueue {
         self.queue.push_back(DeferredAction::Add(object_path))
     }
 
-    /// Push a Remove action onto the back of the queue.
-    pub fn push_remove(&mut self, object_path: Path<'static>) {
-        self.queue.push_back(DeferredAction::Remove(object_path))
+    /// Push Remove actions for a path and all paths that reference it via
+    /// the OpContext.parent field.
+    pub fn push_remove(&mut self, item: &Path<'static>, tree: &Tree<MTFn<TData>, TData>) -> () {
+        for opath in tree.iter() {
```

Comment:
It is preferable to make the recursive call only on the value the algorithm actually recurs on, which in this case is the object path.

Also, it's probably worth making a comment on the algorithmic complexity. At first glance, it looks quadratic, and assuming a large number of object paths, that wouldn't be so great.

Reply:
I have no objections to this suggestion, but am having some trouble getting it to work (i.e. compile). Any chance you could take a stab?

Reply:
For the first part, recursive closure are forbidden in Rust: http://smallcultfollowing.com/babysteps/blog/2013/04/30/the-case-of-the-recurring-closure/. That's a real drawback.

Reply:
Alas, it looks like any attempt might shortly be outlawed. But I'ld still prefer something more like this https://github.com/mulkieran/stratisd/tree/agrover-fix-pool-destroy-for-blockdevs for the following reasons:
1. It makes it explicit that this is the transitive closure of the child relation. That's what the recursion is trying to achieve.
2. It uses filter instead of an if, which I just think is clearer.

Or, if it is not necessary to be recursive, and we think it won't be for the foreseeable future, just don't use recursion.

Reply:
Yeah, `filter` is good here. Ok, will respin based on your version.